Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -35,4 +48,36 @@ void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
     // This is where you can update your variables
     // Hint: There are many different ways that the totalAmpHoursUsed can be updated
     // i.e: Taking a running average of your data values, using most recent data points, etc.
+
+    // The section below is updates the total Amphours used
+    if (dataPoints_ > 0)
+        timeDiff_ = (double)(batteryData_.time.msecsTo(batteryData.time)*0.000000278);
+
+    batteryData_.time = batteryData.time;
+
+    if (dataPoints_ == 0)
+        totalAmpHoursUsed_ = ampHours_;
+    if (dataPoints_ > 0)
+        totalAmpHoursUsed_ += (((batteryData.current + batteryData_.current) /2) * timeDiff_);
+
+    batteryData_.current = batteryData.current;
+
+    dataPoints_++;
+
+    // The section below is for calculating the time it takes to charge or deplete
+    double mSeconds;
+    timeWhenChargedOrDepleted_.setHMS(0,0,0,0);
+
+    if (batteryData_.current < 0)
+    {
+       mSeconds = (totalAmpHoursUsed_/(batteryData_.current*(-1)))*3600000;
```

Comment:
Instead of multiplying by -1, use the absolute value.
Also, the constant `3600000` is what we would call a "Magic Number" as in a number that just kinda shows up with no real explanation. For these numbers, put them in the namespace as a const int, just like in logfilereader.